### PR TITLE
CONSOLE-2425: Support localization of dynamic plugins

### PR DESCRIPTION
**Pull request by jhadvig**
Implementing additional endpoint for handling locales. The endpoint has following format:
`api/locales/<language>/<plugin-locale-file>`
eg. `api/locales/en/olm.json`

When the request hits the endpoint:
1. it will parse out language and plugin's locale file
2. then it will compare the plugin name against the list of enabled plugins
3. if it matches any of the enabled plugins, it will redirect to its service
4. if it doesn't match any of the enabled plugins, use redirect to `loadPath`, which is set to `static/locales/{{lng}}/{{ns}}.json`

Hopefully I understood the desired logic correctly.
@spadgett shall I also update the i18n enhancement ?

PTAL

### CODE REVIEW

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:95]: I wish we were using something like https://github.com/gorilla/mux to handle variables in the path, although that seems like overkill for one endpoint.

I'd suggest simply using query parameters `lng` and `ns` to avoid having to parse the path. We might eventually support multi-loading, and the query parameters would fit better with that (although that could be difficult if we need to proxy some requests).

https://github.com/i18next/i18next-http-backend#backend-options

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:60]: Is this actually making an HTTP request for static files? Can we serve static files directly instead using `http.ServeFile`?

**Review by spadgett — COMMENTED** [review]
Thanks, @jhadvig. We'll want to update the frontend code as well to use the new endpoint.

File: pkg/plugins/handlers.go
Comment (jhadvig) [reply pkg/plugins/handlers.go:60]: we should

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:46]: This isn't a locale. I'd stick to `namespace` as the variable name.

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:45]: Can you open a PR to update the dynamic plugin enhancement with this proposal? Then we can get agreement from stakeholders.

cc @christianvogt @vojtechszocs

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:53]: We should confirm that i18next-http-backend handles redirects. It might be better just to proxy the content here if it's not too hard.

File: pkg/plugins/handlers.go
Comment (jhadvig) [reply pkg/plugins/handlers.go:45]: @spadgett https://github.com/openshift/enhancements/pull/815

File: pkg/plugins/handlers.go
Comment (jhadvig) [reply pkg/plugins/handlers.go:53]: yeah, proxying makes more sense ... updated

**Comment by jhadvig** [comment]
@spadgett comments addressed.

**Comment by jhadvig** [comment]
@spadgett updated the PR per updates in the enhancement

**Comment by jhadvig** [comment]
@spadgett I've addted locales to the dynamic-demo-plugin and updated the two Dockerfiles we have for it.

**Comment by spadgett** [comment]
@vojtechszocs @christianvogt FYI

File: frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json
Comment (vojtechszocs) [inline frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json:2]: Nit: please use 2 spaces for indentation (also in other JSON files).

File: pkg/plugins/handlers.go
Comment (vojtechszocs) [inline pkg/plugins/handlers.go:40]:
```suggestion
	// In case of console app or static plugins, the localization namespace should contain name of the json file that contains the localized strings,
```

File: pkg/plugins/handlers.go
Comment (vojtechszocs) [inline pkg/plugins/handlers.go:41]:
```suggestion
	// since various console packages have different names for the localization file.
```

File: pkg/plugins/handlers.go
Comment (vojtechszocs) [inline pkg/plugins/handlers.go:45]:
```suggestion
	// eg. 'plugin__helm-plugin' will fetch `/locales/{lang}/helm-plugin.json`
```

**Review by vojtechszocs — CHANGES_REQUESTED** [review]

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:31]: Naming nit: I'd call this `HandleLocales` since it handles both plugin locales and console app locales.

File: frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json
Comment (spadgett) [inline frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json:2]: Is this used anywhere in the demo plugin?

I would only add the JSON files if we're actually using it somewhere in the demo plugin to give a real example.

Comment (jhadvig) [reply frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json:2]: no its not, in that case I can just put there and empty json

Comment (vojtechszocs) [reply frontend/dynamic-demo-plugin/locales/en/dynamic-demo-plugin.json:2]: We can have a follow-up PR that illustrates localization usage in dynamic demo plugin.

There should be two examples:
- localizing strings in `console-extensions.json`
- localizing strings in a plain function
- localizing strings in a React component

Once this PR is merged and the conventions are settled, I can write that PR.

File: pkg/plugins/handlers.go
Comment (vojtechszocs) [reply pkg/plugins/handlers.go:52]: On the plus side, this would prevent e.g. `plugin__kubevirt-plugin.json` file names :smiley: 

Having i18n namespace match the file name sounds pretty straight-forward, sounds OK to me.

Comment (jhadvig) [reply pkg/plugins/handlers.go:52]: @spadgett I trim the `plugin__` prefix to get the plugin service endpoint form the map of all the active plugins, where we are mapping the plugin name to its service endpoint. Based on the conversation I dont think ConsolePlugin resource name will contain the prefix, thats why Im trimming it to get the real name.

**Comment by jhadvig** [comment]
@spadgett @vojtechszocs I've rebased the PR and addressed comments. Also remove the i18n changes in the dynamic plugin since we will address those as follow up

PTAL

**Comment by spadgett** [comment]
Test failure looks like it might be related.

> Error: Failed: Missing i18n key "YAML" in namespace "public" and language "en."

**Comment by spadgett** [comment]
browser.log has a large amount of missing key errors:

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/9196/pull-ci-openshift-console-master-e2e-gcp-console/1415975376487714816/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/browser.log

**Comment by spadgett** [comment]
Looking deeper, I see the files are failing to load.

```
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=ceph-storage-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=console-shared - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=devconsole - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=container-security - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=console-app - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=git-service - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=gitops-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=helm-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=insights-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=kubevirt-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=knative-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=lso-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=metal3-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=notification-drawer - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=olm - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=public - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=pipelines-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=rhoas-plugin - Failed to load resource: the server responded with a status of 404 (Not Found)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/locales/resource.json?lng=en&ns=topology - Failed to load resource: the server responded with a status of 404 (Not Found)
```

**Comment by jhadvig** [comment]
Hmm wondering if thats related to the PR since there are also other errors

```
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/api/console/knative-event-sources - Failed to load resource: the server responded with a status of 401 (Unauthorized)
[WARNING] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js 0:380641 "Error fetching CRDs for dynamic event sources" Error: Unauthorized
    at https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:123823
    at Generator.next (<anonymous>)
    at https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:123209
    at new Promise (<anonymous>)
    at s (https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:122957)
    at u (https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:123309)
    at https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/main-chunk-ab5505283ddc025dfa57.min.js:1:124936
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/api/console/knative-channels - Failed to load resource: the server responded with a status of 401 (Unauthorized)
[SEVERE] https://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/static/vendors~main-chunk-d7eaff6a7eb98f5b246e.min.js 87 WebSocket connection to 'wss://console-openshift-console.apps.ci-op-cnxvm8dw-75d12.**********************/api/graphql' failed: HTTP Authentication failed; no valid credentials available
```

**Comment by spadgett** [comment]
@jhadvig Those are expected on initial console load since you're not logged in yet.

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:31]: These are resources, not locales.

```suggestion
func (p *PluginsHandler) HandleI18nResources(w http.ResponseWriter, r *http.Request) {
```

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:45]: I think this is the important bit, so I'd trim the comment to just this:

```suggestion
	// In case of the dynamic plugins, the namespace should contain name of the plugin prefixed with 'plugin__' prefix.
	// eg. 'plugin__helm' will fetch `locales/{lang}/plugin__helm.json` from the plugin service
```

File: pkg/plugins/handlers.go
Comment (spadgett) [inline pkg/plugins/handlers.go:46]: We should return 400 bad request if `lng` or `ns` are missing.

**Review by spadgett — APPROVED** [review]
/lgtm
/hold for approvals
/assign @yapei @ahardin-rh @sferich888

**Comment by yapei** [comment]
tested the changes and it is working well
/label qe-approved

**Pull request merged**
